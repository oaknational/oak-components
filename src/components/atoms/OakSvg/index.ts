export { OakSvg } from "./OakSvg";
export type { OakSvgProps, OakSvgNames } from "./OakSvg";
