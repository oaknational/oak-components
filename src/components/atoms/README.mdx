# Atoms

## Rules for components in this folder

- Can import other atoms
- Imports from molecules and organisms are forbidden!
- Generic
- Unstyled
- Avoid embedded logic and state
- No sub-components
  (eg. Box)
