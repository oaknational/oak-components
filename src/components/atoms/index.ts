export * from "./OakBox";
export * from "./OakFlex";
export * from "./OakSpan";
export * from "./OakForm";
export * from "./OakImage";
export * from "./OakTypography";
export * from "./OakHeading";
export * from "./OakP";
export * from "./OakUL";
export * from "./OakOL";
export * from "./OakLI";
export * from "./OakLabel";
export * from "./OakIcon";
export * from "./OakGrid";
export * from "./OakGridArea";
export * from "./OakThemeProvider";
export * from "./OakMaxWidth";
export * from "./OakCloudinaryImage";
export * from "./OakKbd";
export * from "./OakGlobalStyle";
export * from "./OakAnchorTarget";
export * from "./OakFieldSet";
