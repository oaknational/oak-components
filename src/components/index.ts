export * from "./base";
// export * from "./ui"; // uncomment this line once there are components in this folder
// export * from "./integrated"; // uncomment this line once there are components in this folder
