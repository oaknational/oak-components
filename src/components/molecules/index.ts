export * from "./OakFieldError";
export * from "./OakLoadingSpinner";
export * from "./OakPrimaryButton";
export * from "./OakPrimaryInvertedButton";
export * from "./OakSecondaryButton";
export * from "./OakTertiaryButton";
export * from "./OakRadioButton";
export * from "./OakRadioGroup";
export * from "./OakCheckBox";
export * from "./OakTextInput";
export * from "./OakRoundIcon";
export * from "./OakTooltip";
export * from "./OakBackLink";
export * from "./OakBulletList";
export * from "./OakHandDrawnCard";
export * from "./OakHandDrawnCardWithIcon";
export * from "./OakHandDrawnHR";
export * from "./OakPromoTag";
export * from "./OakLessonInfoCard";
export * from "./OakCollapsibleContent";
export * from "./OakLink";
export * from "./OakSecondaryLink";
export * from "./OakDragAndDropInstructions";
export * from "./OakDraggable";
export * from "./OakDroppable";
export * from "./OakDraggableFeedback";
export * from "./OakOutlineAccordion";
export * from "./OakAccordion";
export * from "./OakModal";
export * from "./OakModalCenter";
export * from "./OakCardWithHandDrawnBorder";
export * from "./OakSecondaryButtonAsRadio";
export * from "./OakButtonAsRadioGroup";
