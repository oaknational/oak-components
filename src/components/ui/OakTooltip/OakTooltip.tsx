import React from "react";

import {
  InternalTooltip,
  InternalTooltipProps,
} from "@/components/base/InternalTooltip/InternalTooltip";

/**
 * A tooltip with oven-ready styling
 */
export const OakTooltip = ({
  ...props
}: Pick<InternalTooltipProps, "isOpen" | "tooltip" | "children">) => {
  return (
    <InternalTooltip
      $background="bg-decorative5-main"
      $color="text-primary"
      $pv="inner-padding-m"
      $ph="inner-padding-xl"
      $font="heading-light-7"
      $borderRadius="border-radius-m2"
      {...props}
    />
  );
};
