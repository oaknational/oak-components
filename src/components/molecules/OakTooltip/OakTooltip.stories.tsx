import React from "react";
import { Meta, StoryObj } from "@storybook/react";

import { OakTertiaryButton } from "../OakTertiaryButton";

import { OakTooltip } from "./OakTooltip";

import { OakBox } from "@/components/atoms";

const meta: Meta<typeof OakTooltip> = {
  component: OakTooltip,
  tags: ["autodocs"],
  title: "components/molecules/OakTooltip",
  argTypes: {
    tooltip: {
      control: {
        type: "text",
      },
    },
  },
  parameters: {
    controls: {
      include: ["isOpen", "tooltip"],
    },
  },
  decorators: [
    (Story) => (
      <OakBox $mt="space-between-xxl">
        <Story />
      </OakBox>
    ),
  ],
  args: {
    isOpen: true,
    tooltip: "Hello there",
    children: <OakTertiaryButton>Target</OakTertiaryButton>,
  },
};
export default meta;

type Story = StoryObj<typeof OakTooltip>;

export const Default: Story = {
  render: (args) => <OakTooltip {...args} />,
};
