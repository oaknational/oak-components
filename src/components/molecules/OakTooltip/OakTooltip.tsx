import React, {
  ReactElement,
  ReactNode,
  useLayoutEffect,
  useState,
} from "react";
import { createPortal } from "react-dom";

import {
  InternalTooltip,
  InternalTooltipProps,
} from "@/components/atoms/InternalTooltip/InternalTooltip";
import { OakBox, OakBoxProps } from "@/components/atoms";
import { parseSpacing } from "@/styles/helpers/parseSpacing";

export type OakTooltipProps = Pick<InternalTooltipProps, "tooltipPosition"> & {
  /**
   * The target element that triggers the tooltip
   */
  children: ReactElement;
  /**
   * The content of the tooltip
   */
  tooltip: ReactNode;
  /**
   * Whether the tooltip is open or not
   */
  isOpen?: boolean;
  /**
   * The DOM container to render the tooltip portal into
   *
   * @default document.body
   */
  domContainer?: Element;
  id: string;
};

/**
 * A tooltip with oven-ready styling and positioning.
 */
export const OakTooltip = ({
  tooltipPosition,
  children,
  tooltip,
  isOpen,
  domContainer,
  id,
  ...props
}: OakTooltipProps) => {
  const [targetElement, setTargetElement] = useState<Element | null>(null);
  const [isIntersecting, setIsIntersecting] = useState(true);
  const [domContainerState, setDomContainerState] = useState<
    Element | undefined
  >(domContainer);
  const isVisible = isOpen && isIntersecting;

  /**
   * The overlay is positioned on top of the target element in a portal.
   * It tracks the target's size and position.
   *
   * we use it to position the tooltip relative to the target element without interfering
   * with the page layout
   */
  const [overlayStyle, setOverlayStyle] = useState<{
    top: number;
    left: number;
    width: number;
    height: number;
  }>();
  const squaredCornerRadiusProp: keyof InternalTooltipProps = (() => {
    switch (tooltipPosition) {
      case "bottom-right":
        return "$btrr";
      case "top-right":
        return "$bbrr";
      case "top-left":
        return "$bblr";
      default:
        return "$btlr";
    }
  })();
  const borderRadiusProps: Partial<InternalTooltipProps> = {
    $borderRadius: "border-radius-m",
    [squaredCornerRadiusProp]: "border-radius-square",
  };

  useLayoutEffect(() => {
    if (!targetElement) {
      return;
    }
    let ticking = false;
    const updateOverlayStyle = () => {
      if (!ticking) {
        requestAnimationFrame(() => {
          ticking = false;
          const rect = targetElement.getBoundingClientRect();

          setOverlayStyle({
            top: rect.top,
            left: rect.left,
            width: rect.width,
            height: rect.height,
          });
        });
        ticking = true;
      }
    };

    // We use an intersection observer to detect when the target element is no longer visible
    const intersection = new IntersectionObserver((entries) => {
      setIsIntersecting(entries.every((entry) => entry.isIntersecting));
    });
    intersection.observe(targetElement);

    // We use a resize observer to detect when the target element's size changes
    const resize = new ResizeObserver(updateOverlayStyle);
    resize.observe(targetElement);

    // Update the overlay position on scroll and resize
    window.addEventListener("scroll", updateOverlayStyle, true);
    window.addEventListener("resize", updateOverlayStyle);

    return () => {
      resize.disconnect();
      intersection.disconnect();
      window.removeEventListener("scroll", updateOverlayStyle, true);
      window.removeEventListener("resize", updateOverlayStyle);
    };
  }, [targetElement]);

  useLayoutEffect(() => {
    if (domContainerState) {
      return;
    }
    setDomContainerState(document.body);
  }, [domContainerState]);

  return (
    <>
      {domContainerState &&
        createPortal(
          isVisible && (
            <OakBox
              $position="fixed"
              style={overlayStyle}
              $pointerEvents="none"
              $zIndex="modal-dialog"
            >
              <OakBox
                $width="fit-content"
                $height="fit-content"
                $position="absolute"
                {...getTooltipPositionProps(tooltipPosition)}
              >
                <InternalTooltip
                  $background="bg-decorative5-main"
                  $color="text-primary"
                  $pv="inner-padding-m"
                  $ph="inner-padding-xl"
                  $font="heading-light-7"
                  tooltipPosition={tooltipPosition}
                  {...props}
                  {...borderRadiusProps}
                  aria-expanded={isVisible}
                  id={id}
                >
                  {tooltip}
                </InternalTooltip>
              </OakBox>
            </OakBox>
          ),
          domContainerState,
        )}
      <div
        ref={(domElement) => {
          setTargetElement(domElement?.firstElementChild ?? null);
        }}
        style={{ display: "contents" }}
      >
        {children}
      </div>
    </>
  );
};

function getTooltipPositionProps(
  tooltipPosition: OakTooltipProps["tooltipPosition"],
) {
  const props: Partial<OakBoxProps> = {};

  switch (tooltipPosition) {
    case "top-left":
    case "top-right":
      props.$top = "space-between-none";
      props.$transform = `translateY(calc(-100% - ${parseSpacing(
        "space-between-s",
      )}))`;
      break;
    default:
      props.$bottom = "space-between-none";
      props.$transform = `translateY(calc(100% + ${parseSpacing(
        "space-between-s",
      )}))`;
      break;
  }

  switch (tooltipPosition) {
    case "top-left":
    case "bottom-left":
      props.$left = "space-between-none";
      break;
    default:
      props.$right = "space-between-none";
      break;
  }

  return props;
}
