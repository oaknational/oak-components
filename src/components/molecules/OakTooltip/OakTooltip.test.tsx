import { create } from "react-test-renderer";
import React, { ReactNode } from "react";
import "@testing-library/jest-dom";

import { OakTooltip } from "./OakTooltip";

import renderWithTheme from "@/test-helpers/renderWithTheme";
import { oakDefaultTheme } from "@/styles";
import { OakThemeProvider } from "@/components/atoms";
import {
  installMockIntersectionObserver,
  installMockResizeObserver,
} from "@/test-helpers";

jest.mock("react-dom", () => {
  return {
    ...jest.requireActual("react-dom"),
    createPortal: (node: ReactNode) => node,
  };
});

installMockResizeObserver();
installMockIntersectionObserver();

describe(OakTooltip, () => {
  it("matches snapshot", () => {
    const tree = create(
      <OakThemeProvider theme={oakDefaultTheme}>
        <OakTooltip tooltip="Hello there" isOpen>
          <div>Trigger!</div>
        </OakTooltip>
      </OakThemeProvider>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("renders", () => {
    const { getByRole } = renderWithTheme(
      <OakTooltip tooltip="Hello there" isOpen>
        <div>Trigger!</div>
      </OakTooltip>,
    );

    expect(getByRole("tooltip")).toBeInTheDocument();
  });
});
