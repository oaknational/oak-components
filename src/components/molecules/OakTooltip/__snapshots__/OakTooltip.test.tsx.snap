// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OakTooltip matches snapshot 1`] = `
.c0 {
  position: fixed;
  pointer-events: none;
  font-family: --var(google-font),Lexend,sans-serif;
  z-index: 300;
}

.c1 {
  position: absolute;
  right: 0rem;
  bottom: 0rem;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  height: -webkit-fit-content;
  height: -moz-fit-content;
  height: fit-content;
  -webkit-transform: translateY(calc(100% + 1rem));
  -ms-transform: translateY(calc(100% + 1rem));
  transform: translateY(calc(100% + 1rem));
  font-family: --var(google-font),Lexend,sans-serif;
}

.c2 {
  position: relative;
  max-width: 22.5rem;
  padding-left: 1.5rem;
  padding-right: 1.5rem;
  padding-top: 1rem;
  padding-bottom: 1rem;
  color: #222222;
  background: #ffe555;
  border-radius: 0.375rem;
  border-top-left-radius: 0rem;
  font-family: --var(google-font),Lexend,sans-serif;
  font-weight: 400;
  font-size: 1rem;
  line-height: 1.25rem;
  -webkit-letter-spacing: 0.0115rem;
  -moz-letter-spacing: 0.0115rem;
  -ms-letter-spacing: 0.0115rem;
  letter-spacing: 0.0115rem;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c4 {
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  pointer-events: none;
}

.c5 {
  position: absolute;
  top: -1rem;
  left: 0rem;
  -webkit-transform: scaleY(-1);
  -ms-transform: scaleY(-1);
  transform: scaleY(-1);
  fill: #ffe555;
}

@media (min-width:750px) {
  .c2 {
    max-width: 40rem;
  }
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        aria-expanded="true"
        class="c2 c3 c4"
        data-rac="true"
        id="tooltip"
        role="tooltip"
      >
        Hello there
        <svg
          class="c5"
          data-testid="tooltip-arrow"
          height="1rem"
          width="1rem"
        >
          <path
            d="M0 0H16L8 8L0 16V0Z"
          />
        </svg>
      </div>
    </div>
  </div>
  <div
    style="display: contents;"
  >
    <div>
      Trigger!
    </div>
  </div>
</div>
`;
